polygeist/lib/polygeist/Passes/ConvertToLLVMABI.cpp

```diff
@@ -0,0 +1,123 @@
+
+namespace {
+
+class GPUFuncLowering : public OpRewritePattern<gpu::GPUFuncOp> {
+public:
+  using OpRewritePattern<gpu::GPUFuncOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(gpu::GPUFuncOp op,
+                                PatternRewriter &rewriter) const override {
+    assert(op->getAttr("llvm.cconv") ==
+               mlir::LLVM::CConvAttr::get(op.getContext(),
+                                          LLVM::cconv::CConv::SPIR_KERNEL) &&
+           "Expecting SYCL Kernel");
+    LLVM_DEBUG(llvm::dbgs() << "ConvertToLLVMABIPass: GPUFuncLowering: "
+                            << op.getName() << "\n");
+
+    // Notify MLIR we're updating the function in place
+    rewriter.startRootUpdate(op);
+
+    FunctionType funcTy = op.getFunctionType();
+    TypeConverter::SignatureConversion conversion(funcTy.getNumInputs());
+
+    rewriter.setInsertionPointToStart(&op.getBody().front());
+    Region &funcBody = op.getBody();
+    SmallVector<BlockArgument, 8> Args(funcBody.getArguments().begin(),
+                                       funcBody.getArguments().end());
+    for (auto &en : llvm::enumerate(Args)) {
+      Value arg = en.value();
+      auto MT = arg.getType().dyn_cast<MemRefType>();
+      if (!MT) {
+        conversion.addInputs(en.index(), {arg.getType()});
+        continue;
+      }
+
+      LLVM_DEBUG(llvm::dbgs() << "  Replace argument " << en.index() << ": \""
+                              << arg << "\"");
+      Type PT = LLVM::LLVMPointerType::get(MT.getElementType(),
+                                           MT.getMemorySpaceAsInt());
+      Value newArg = funcBody.insertArgument(en.index(), PT, arg.getLoc());
+      LLVM_DEBUG(llvm::dbgs() << " -> \"" << newArg << "\"\n");
+      conversion.addInputs(en.index(), {PT});
+
+      auto Ptr2Memref = rewriter.create<polygeist::Pointer2MemrefOp>(
+          arg.getLoc(), MT, newArg);
+
+      arg.replaceAllUsesWith(Ptr2Memref);
+      funcBody.eraseArgument(en.index() + 1);
+    }
+
+    assert(funcTy.getNumResults() == 0 &&
+           "Expecting SYCL kernel to return void");
+
+    auto newFuncTy = FunctionType::get(
+        op.getContext(), conversion.getConvertedTypes(), funcTy.getResults());
+    op->setAttr(FunctionOpInterface::getTypeAttrName(),
+                TypeAttr::get(newFuncTy));
+    LLVM_DEBUG(llvm::dbgs() << "  New FunctionType: " << newFuncTy << "\n");
+
+    // Notify MLIR in place updates are done
+    rewriter.finalizeRootUpdate(op);
+
+    return success();
+  }
+};
+
+struct ConvertToLLVMABIPass final
+    : public mlir::polygeist::ConvertToLLVMABIBase<ConvertToLLVMABIPass> {
+  void runOnOperation() override {
+    ModuleOp m = getOperation();
+
+    RewritePatternSet patterns(&getContext());
+    patterns.add<GPUFuncLowering>(&getContext());
+
+    ConversionTarget target(getContext());
+    target.addDynamicallyLegalOp<gpu::GPUFuncOp>([](gpu::GPUFuncOp op) {
+      return llvm::none_of(op.getFunctionType().getInputs(),
```

Comment:
Check also the return value(s). I know that SPIR kernels must return `void`, however I believe a generic `gpu.func` might return one (or more) values.

Reply:
Added an assert for now to ensure `op` returns `void`. Will change this when adding the return type handling.